Add myopic option to config, snakefile and the necessary additional scripts

martavp opened the pull request:
This pull request converts the crude myopic model used in [pypsa-eur-sec-30-path](https://github.com/martavp/pypsa-eur-sec-30-pathl) into code more compatible with pypsa-eur-sec.

Main changes already implemented are:

* conditional entering on myopic code based on config file (if config["foresight"] == "myopic")

* existing onwind and offwind capacity are now read from IRENA database, similarly to solar capacities. Previously we were using thewindpower.net database which is not open

*coding standard of PyPSA-Eur-Sec

*no technologies hardcoded in scripts

* for every technology, different generators are implemented for every time step, capacities installed in a previous timestep are saved as e.g., solar-2030, following the convention for perfect foresight.

* existing heating capacities (installed before base year) are also added

* Land use is included as a global constraint summing the capacities of a certain technology installed in different years

The model is working now for the one-one-per-country network, but not when a larger number of nodes are included. 
To that end, the best strategy will be to create a powerplants_cluster.csv file that can be used later. 
We need to think about how to split existing wind and solar capacity (reported at a national level) when more than one node per country are used.

nworbmot commented:
Thanks @martavp, this looks excellent! Thanks for documenting everything so clearly and keeping the code so clean and easy to read. I will look at this today.

martavp commented:
Perfect, @nworbmot. I just pushed three additional small commits.
I'm done now and the code is ready to be reviewed.

nworbmot commented on scripts/add_brownfield.py:
```diff
@@ -0,0 +1,183 @@
+override_component_attrs["Link"].loc["p3"] = ["series","MW",0.,"3rd bus output","Output"]
+
+
+def prepare_costs():
```
Do we need this function prepare_costs() defined here if we already imported it from add_existing_baseyear?

martavp replied:
We can import it from add_existing_baseyear. I have a meeting now but I will include it tomorrow

martavp replied:
This is done in [this commit](https://github.com/PyPSA/pypsa-eur-sec/pull/63/commits/7d1500604fd1027a7d8952daa4cc40c58444ff66). 

prepare_costs() function has been modified to include the 2 levels multindex files produced in the t[echnology_data repository](https://github.com/lisazeyen/technology_data/tree/master/outputs).

hydrogen underground storage is replaced by gas storage until the name is updated in the technolgy_data repository

nworbmot commented on scripts/add_brownfield.py:
```diff
@@ -0,0 +1,183 @@
+    n_p.mremove("Generator", [index for index in n_p.generators.index.to_list() if '<'+snakemake.config['scenario']['planning_horizons'][0] in index])
+    n_p.mremove("Generator", [index for index in n_p.generators.index.to_list() if 'ror' in index])
+        
+    n_p.generators.index=np.where(n_p.generators.index.str[-4:].isin(previous_timesteps)==False,
```
I don't 100% understand what's going on here. So if it doesn't have a -year suffix in previous_timesteps, then we just assume it's from the previous year? What generators fall into this category? Don't we need to take account of the asset lifetime here? E.g. 20-year-old asset built in 2020 will not be there in 2050.

martavp replied:
I have improved the code and the commented lines that explain it:
1.	If it doesn’t have a -year suffix in previous_timesteps, we assume that it was optimized in the previous time_step
2.	We remove those generators whose installationYear + lifetime < year
3.	We rename capacities optimized in the previous timestep by adding the suffix eg. -2030 and add them to the network.

All the generators are included into this category (except ror which are removed since their capacities are exogenously fixed).

So far, I’m assuming a “default” lifetime (included in the config file). 
What would be the best way of recover technology names? so that we can read costs.at[technology, ‘lifetime’].

I was considering using the generators name or carrier but it is not direct. Maybe I should just wait until the attribute “lifetime” is added (I understood that this was the plan for the perfect foresight approach).

nworbmot commented on scripts/add_brownfield.py:
```diff
@@ -0,0 +1,183 @@
+    
+    baseyear=snakemake.config['scenario']["planning_horizons"][0]
+    
+    add_power_capacities_installed_before_baseyear(n, year, baseyear, costs) # only the capacities with YearDecomissioning < year are added
```
YearDecomissioning > year?

martavp replied:
Yep, corrected in https://github.com/PyPSA/pypsa-eur-sec/pull/63/commits/a4c22a29a43f63e52d879267a7f26a5d27b5faa5

nworbmot commented on scripts/add_existing_baseyear.py:
```diff
@@ -0,0 +1,394 @@
+    
+    for node in nodes:
+        #if a country has more than one node, selects the first one
+        bus_selected=[bus for bus in n.buses.index[n.buses.carrier == "AC"] if bus[0:2]==node][0]  
```
We need to fix this soonish so that it attaches generators to the correct node, or it should go on a TODO list. Powerplantmatching has the coordinates.

martavp replied:
For those plants included in powerplantmatching should be easy, for existing wind and solar capacities we need to decide how to split among nodes in a country

nworbmot commented on scripts/add_existing_baseyear.py:
```diff
@@ -0,0 +1,394 @@
+               carrier=name + " resistive heater",
+               efficiency=costs.at[name_type + ' resistive heater','efficiency'],
+               capital_cost=costs.at[name_type + ' resistive heater','efficiency']*costs.at[name_type + ' resistive heater','fixed'],
+               p_nom=0.5*df['{} resistive heater'.format(heat_type)][nodes[name]]*max(0,(2045-int(year))/25))
```
The years 2045 and lifetime assumptions should go in the config.yaml.

martavp replied:
Added in https://github.com/PyPSA/pypsa-eur-sec/pull/63/commits/2b3dc60f11179d4efd3cd3a4f11f86fe679dab9c and https://github.com/PyPSA/pypsa-eur-sec/pull/63/commits/a4c22a29a43f63e52d879267a7f26a5d27b5faa5

nworbmot commented on scripts/solve_network.py:
```diff
@@ -155,10 +155,21 @@ def add_chp_constraints(n):
     #add_eps_storage_constraint(n)
     add_chp_constraints(n)
+    add_land_use_constraint(n)
     add_battery_constraints(n)
```
Won't add_land_use break for the overnight foresight case? Maybe make it conditional on config["foresight"]?

martavp replied:
Yes, you are right. Implemented in https://github.com/PyPSA/pypsa-eur-sec/pull/63/commits/dedabe695ed7da305e05d86c05a5938414b18040

martavp commented:
>     * I will do a release of PyPSA-Eur-Sec as it currently is in the master branch, e.g. a 0.1.0 release, so then we can have instructions for people upgrading from 0.1.0 to 0.2.0 with myopic foresight option.

Perfect.
 
>     * To avoid duplicating the snakemake rules like prepare_sector_network, we can add the co2_budget_name and planning_horizons wildcards to the "overnight" case with e.g. defaults ["vanilla"] and [2050] which are then ignored by the scripts. Could you implement this change so that no rules are unnecessarily copied in the Snakefile?

Done in https://github.com/PyPSA/pypsa-eur-sec/pull/63/commits/b23b7a143c4a521e5f39150f72c1514308d0bf1a

>     * Maybe rename "prenetworks-bf" to "prenetworks-brownfield" since it's clearer?

Done also in https://github.com/PyPSA/pypsa-eur-sec/pull/63/commits/b23b7a143c4a521e5f39150f72c1514308d0bf1a

>     * What is the origin and licence for the imported data on existing renewables and heating technologies?

The data on solar, onwind and offwind capacities comes from IRENA.
I have added a commented [link to the script] (https://github.com/martavp/pypsa-eur-sec/blob/201e4d2f5732d637d11d3683608d87783c439827/scripts/add_existing_baseyear.py#L108)

I read the data from the [statistics tool](https://www.irena.org/Statistics/Download-Data)
This gathers data included in IRENA report, such [this from 2020] (https://www.irena.org/-/media/Files/IRENA/Agency/Publication/2020/Mar/IRENA_RE_Capacity_Statistics_2020.pdf)
In page 2 they have the following Copyright statement:
Copyright © IRENA 2020Unless otherwise stated, material in this publication may be freely used, shared, copied, reproduced, printed and/or stored, provided that appropriate acknowledgement is given of IRENA as the source and copyright holder. Material in this publication that is attributed to third parties may be subject to separate terms of use and restrictions, and appropriate permissions from these third parties may need to be secured before any use of such material.

The data on existing heating capacities comes for [this study]
(https://ec.europa.eu/energy/studies/mapping-and-analyses-current-and-future-2020-2030-heatingcooling-fuel-deployment_en?redir=1)
I have added a commented [link to the script](https://github.com/martavp/pypsa-eur-sec/blob/201e4d2f5732d637d11d3683608d87783c439827/scripts/add_existing_baseyear.py#L225)
I cannot find any license attached to the dataset or the reports. I can write an email to the main authors and ask them to add an open license to the report and the dataset.

>     * I missed where the conventional power plants are read in?
Here : https://github.com/martavp/pypsa-eur-sec/blob/201e4d2f5732d637d11d3683608d87783c439827/scripts/add_existing_baseyear.py#L166

>     * I made some more specific comments at points in the code (e.g. new extra_functionality constraint on land use breaks solve_network for the overnight case, and 2045 is hard-coded in the add_existing_baseyear)

Done.
 
>     * What do users need to do to get the costs-2040.csv files?
These are the outcomes of the technolgy_data repository https://github.com/lisazeyen/technology_data/tree/master/outputs

nworbmot commented:
Brilliant, thanks @martavp for taking care of this so quickly! Looks like we'll soon be done! I'll get on with the release and starting the documentation. Tomorrow @lisazeyen and I will sort out getting the cost database opened up and in the main PyPSA directory.

nworbmot commented:
Release is done! To avoid using deductions from the generator/link/line names for build years and lifetimes, I think we should already now introduce "lifetime" (float) and "build_year" (int?) attributes everywhere, i.e. new columns in `network.generators` etc. New attributes can be introduced using the `network.madd` (for `network.add`, you need to add the new column to the pandas.DataFrame `network.generators` first).

martavp commented:
> I think we should already now introduce "lifetime" (float) and "build_year"

I agree. I will introduce the attributes “lifetime” and “build_year” in the myopic code. The decomissioning rules will be much easier. 
I can take care of this early next week.

martavp commented:
> 
> 
> Release is done! To avoid using deductions from the generator/link/line names for build years and lifetimes, I think we should already now introduce "lifetime" (float) and "build_year" (int?) attributes everywhere, i.e. new columns in `network.generators` etc. New attributes can be introduced using the `network.madd` (for `network.add`, you need to add the new column to the pandas.DataFrame `network.generators` first).

This is done in https://github.com/PyPSA/pypsa-eur-sec/pull/63/commits/fa0b88a9b101873b14e578b5e7a4a754fd2803b0

For the (power & heating) capacities installed prior to base year, their DecomissioningYear is evaluated before aggregating them to the network.  Hence, it does not make sense to add a lifetime and build_year attribution.

For those capacities installed in a previous time steps, the script add_brownfield adds the attributes base_year and lifetime. 
For the lifetime still a default lifetime it used. The specific lifetime associated with every technology could be added using the name/carrier of the generator/link/store to identify the technology.

nworbmot commented:
I've tested the code now and can confirm that, after a fixing a few issues (most important was the bug with `mincount=1` in the costs processing - please check old pathway code that this bug is not there, since it causes discount rate to default to zero!) the overnight still works great, and that the myopic is nearly there. Still to do that I see:
- [ ] I think the costs for wind and solar, including the AC and DC connection costs for offshore, are still being set by PyPSA-Eur and not updated with the costs_20xx.
- [ ] Fix the mapping of conventional power plants to buses (currently assigned to first bus in the correct country) - can use bus regions.
- [ ] The existing capacities of wind and solar in Germany in 2019 are too low (40 GW for solar versus 52 GW in IRENA, 35 GW for onshore wind versus 45 GW in IRENA). Bug?
- [ ] Somehow deal with the existing energy carriers for transport and industry, or warn that myopic only adds existing electricity and heat.
- [ ] The extra functionality `add_land_constraints` in `solve_network.py` is causing an error with infinity on the right hand side - I've had to comment it out.
- [ ] The pattern bus.index[0:5] to find the location won't work for "DE0 15" etc. - should store "location" on all buses to make this easier (a general issue for PyPSA-Eur-Sec).
- [ ] Define conventional carriers (coal, oil, gas, uranium) in config.yaml and have them added in a more uniform way.
- [ ] Write documentation and guidelines for myopic code, including rough code structure and what works and what doesn't.

nworbmot commented:
I'm taking care of the nodal distribution of powerplants and the `copy_config` rule this week, in case anyone else is working in parallel on this.

martavp commented:
Thanks, Tom. I have been busy after the holidays but I have reserved some time tomorrow and Wednesday. I will start with updating pypsa-eur costs for wind and solar, checking the possible bugs that you mentioned and reviewing the add_land_constraint rule.

FabianHofmann commented:
@martavp considering the `add_land_constraint`: I've just proposed an very similar implementation in pypsa, see https://github.com/PyPSA/PyPSA/pull/197. The idea is to limit the combined capacity of generators of the same carrier at the same bus through values in n.buses['nom_max_{carrier}'], this might help you here.

martavp commented:
(most important was the bug with `mincount=1` in the costs processing - please check old pathway code that this bug is not there, since it causes discount rate to default to zero!) 

I checked that this was not a bug in the old pathway code since we were using pandas < 0.22

>     * [x]  The existing capacities of wind and solar in Germany in 2019 are too low (40 GW for solar versus 52 GW in IRENA, 35 GW for onshore wind versus 45 GW in IRENA). Bug?

My mistake, I forgot the equal in “lower and equal” when grouping the existing technologies by grouping year. But your implementation with digitize works fine.

nworbmot commented:
I've made some simplifications to the myopic code - now the year being optimized (e.g. 2020 in first solving run) is added to the newly optimized assets as a `build_year` and in the name in `add_existing_baseyear.py` and similarly for new assets in `add_brownfield.py`. This allows a radical simplification of the `add_brownfield.py` code to a few lines, and prepares us better for the perfect foresight later. I'm just testing the code and will push the changes today/tomorrow.

nworbmot commented:
My big changes are now done! Hopefully the `add_brownfield.py` is much simpler now. @martavp please check I haven't done anything wrong. The model ran fine through to 2030, will try to 2050 now.

nworbmot commented:
After fixing some bugs with the CHPs, it works until 2050! I didn't test the updated wind/solar costs yet.
I think the code should be ready to pull into the main repository next week.
Some things we can consider for future development:

- [ ] Assumptions for a split rooftop-utility for the existing solar
- [ ] Squashing assets with p_nom_extendable=False and the same attributes (including time-dependent attributes like p_max_pu), except p_nom and build_year, into a single asset to reduce the number of variables. We'd need a script to squash assets before solving, and unsquash them afterwards. I have a code sketch, but don't want to implement it right now.
- [ ] `config.default.yaml` goes to overnight (since this is the best developed) and we offer a `config.myopic.yaml` alternative.

martavp commented:
> My big changes are now done! Hopefully the `add_brownfield.py` is much simpler now. @martavp please check I haven't done anything wrong. The model ran fine through to 2030, will try to 2050 now.

I went through the changes and everything seems fine. add_brownfield is much simpler now, great!

martavp commented:
>  Assumptions for a split rooftop-utility for the existing solar

I'll look for some statistics to implement this splitting

martavp commented:
> @martavp considering the `add_land_constraint`: I've just proposed an very similar implementation in pypsa, see [PyPSA/PyPSA#197](https://github.com/PyPSA/PyPSA/pull/197). The idea is to limit the combined capacity of generators of the same carrier at the same bus through values in n.buses['nom_max_{carrier}'], this might help you here.


Thanks, @FabianHofmann . A very simple approach is implemented now. I.e., p_nom for already installed capacities are substrated from the p_nom_max of extendable capacities. We should use your suggested cleaner version once it is available in PyPSA

nworbmot commented:
I'm doing some final tests, but I think this is ready to pull into the master branch, triggering a 0.2.0 release!
I won't change any content-ful code, but I may move some things around (e.g. config.default.yaml to overnight, with a separate config.myopic.yaml, remove data/costs.csv, and I may put the brownfield prenetworks in the same folder as the prenetworks for simplicity).
I will also do a 0.1.0 release on the technology_data first.

martavp commented:
Perfect, all those small changes seem reasonable.
